DAOS-16553 client: intercept ucs_init and memset ucs global variable after fork

wiliamhuang opened the pull request:
Intercepts ucs_init(). Query the address of global variable "ucs_async_thread_global_context" in libucs inside ucs_init() and set up a fork handler to zero "ucs_async_thread_global_context" in child process after fork().

### Steps for the author:

* [ ] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [ ] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [ ] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [ ] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [ ] Gatekeeper requested (daos-gatekeeper added as a reviewer).

wiliamhuang committed:
DAOS-16553 client: intercept ucs_init and memset ucs global variable after fork

Features: pil4dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/1/execution/node/1049/log

wiliamhuang committed:
add code to debug the issue in CI

Skip-unit-tests: true
Skip-func-hw-test: true
Test-tag: test_dfuse_daos_build_wt,pil4dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
empty commit to trigger rebuild

Skip-unit-tests: true
Skip-func-hw-test: true
Test-tag: pil4dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
print more info for debugging

Skip-unit-tests: true
Skip-func-hw-test: true
Test-tag: pil4dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

daosbuild3 commented:
Test stage Build on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/5/execution/node/310/log

daosbuild3 commented:
Test stage Build on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/6/execution/node/309/log

daosbuild3 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/6/execution/node/304/log

daosbuild3 commented:
Test stage Build on Leap 15.5 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/6/execution/node/405/log

daosbuild3 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/7/execution/node/302/log

daosbuild3 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/8/execution/node/311/log

wiliamhuang committed:
empty commit to trigger rebuild

Skip-unit-tests: true
Skip-func-hw-test: true
Test-tag: pil4dfs
Allow-unstable-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

daosbuild3 commented:
Test stage Build on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/8/execution/node/405/log

wiliamhuang committed:
fix compiling issue in CI

Skip-unit-tests: true
Skip-func-hw-test: true
Test-tag: pil4dfs
Allow-unstable-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/10/execution/node/720/log

wiliamhuang committed:
return if dlsym() fails to find ucs_init

Allow-unstable-test: true
Features: pil4dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

frostedcmos commented on src/client/dfuse/pil4dfs/hook.c:
```diff
@@ -1186,3 +1187,101 @@ query_all_org_func_addr(void)
+
+	rc = stat(dl_info.dli_fname, &file_stat);
+	if (rc == -1) {
+		D_DEBUG(DB_ANY, "Fail to query stat of file %s", dl_info.dli_fname);
```
should these not be D_ERRORs instead? Same for few other D_DEBUG prints before in error paths

frostedcmos commented on src/client/dfuse/pil4dfs/hook.c:
```diff
@@ -1186,3 +1187,101 @@ query_all_org_func_addr(void)
+			}
+			num_sym = sections[i].sh_size / sections[i].sh_entsize;
+
+			/* i should be larger than 0 here given sections[i].sh_type == SHT_SYMTAB.
```
sounds like a good place to assert that i > 0 and < header->e_shnum - 2

wiliamhuang committed:
Use D_ERROR instead of D_DEBUG
Avoid possible boundary issue

Allow-unstable-test: true
Features: pil4dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/13/execution/node/1335/log

mchaarawi commented:
please add description to the PR

wiliamhuang committed:
empty commit to trigger rebuild

Test-tag: test_pil4dfs_vs_dfs
Test-tag-hw-medium-ucx-provider: test_pil4dfs_vs_dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang commented:
> we should also run the tests with the UCX provider, otherwise testing is useless in the PR

Just restarted PR with,
Test-tag: test_pil4dfs_vs_dfs
Test-tag-hw-medium-ucx-provider: test_pil4dfs_vs_dfs

daosbuild3 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-17010/17/testReport/

wiliamhuang committed:
empty commit to trigger rebuild

Allow-unstable-test: true
Test-tag: test_pil4dfs_vs_dfs
Test-tag-hw-medium-ucx-provider: test_pil4dfs_vs_dfs

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
empty commit to trigger rebuild

Test-tag-hw-medium-ucx-provider: test_pil4dfs_vs_dfs
Allow-unstable-test: true
Skip-unit-tests: true
RPM-test: false
Skip-fault-injection-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
empty commit to trigger rebuild

Test-tag-hw-medium-ucx-provider: test_pil4dfs_vs_dfs
Allow-unstable-test: true
Skip-unit-tests: true
RPM-test: false
Skip-fault-injection-test: true
Skip-func-test-el8: true
Skip-test-centos-rpms: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
add provider to test_pil4dfs_vs_dfs

Test-tag: test_pil4dfs_vs_dfs
Test-tag-hw-medium-ucx-provider: test_pil4dfs_vs_dfs
Allow-unstable-test: true
Skip-unit-tests: true
RPM-test: false
Skip-fault-injection-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
empty commit to trigger rebuild

Test-provider: ucx+dc_x
Test-tag: test_pil4dfs_vs_dfs
Allow-unstable-test: true
Skip-unit-tests: true
RPM-test: false
Skip-fault-injection-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
remove provider tag for test_pil4dfs_vs_dfs

Test-provider: ucx+dc_x
Test-tag: test_pil4dfs_vs_dfs
Allow-unstable-test: true
Skip-unit-tests: true
RPM-test: false
Skip-fault-injection-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
empty commit to trigger rebuild

Test-provider: ucx+dc_x
Test-tag: test_pil4dfs_vs_dfs
Allow-unstable-test: true
Skip-fault-injection-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
determine module base address and function address

Test-provider: ucx+dc_x
Test-tag: pil4dfs test_pil4dfs_vs_dfs
Allow-unstable-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
fix clang issues

Test-provider: ucx+dc_x
Test-tag: pil4dfs test_pil4dfs_vs_dfs
Allow-unstable-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
fix clang issue and use printf instead of D_ERROR

Test-provider: ucx+dc_x
Test-tag: pil4dfs test_pil4dfs_vs_dfs
Allow-unstable-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

wiliamhuang committed:
Merge branch 'master' into lei/DAOS-16553

Test-provider: ucx+dc_x
Test-tag: pil4dfs test_pil4dfs_vs_dfs
Allow-unstable-test: true

Signed-off-by: Lei Huang <lei.huang@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/28/execution/node/1026/log

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/28/execution/node/1296/log

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17010/28/execution/node/1341/log
